src/client/common/process/interpreterActivation.ts

```diff
@@ -0,0 +1,379 @@
+    ): Promise<NodeJS.ProcessEnv | undefined> {
+        const stopWatch = new StopWatch();
+        const env = await this.apiProvider
+            .getApi()
+            .then((api) => api.getActivatedEnvironmentVariables(resource, interpreter, false));
+
+        const envType = interpreter.envType;
+        sendTelemetryEvent(Telemetry.GetActivatedEnvironmentVariables, stopWatch.elapsedTime, {
+            envType,
+            pythonEnvType: envType,
+            source: 'python',
+            failed: Object.keys(env || {}).length === 0,
+            reason: 'emptyVariables'
+        });
+        // We must get actiavted env variables for Conda env, if not running stuff against conda will not work.
+        // Hence we must log these as errors (so we can see them in jupyter logs).
+        if (!env && envType === EnvironmentType.Conda) {
+            traceError(`Failed to get activated conda env variables for ${getDisplayPath(interpreter?.path)}`);
+        }
+        return env;
+    }
+    @traceDecorators.verbose(
+        'Getting activated env variables ourselves',
+        TraceOptions.BeforeCall | TraceOptions.Arguments
+    )
+    private async getActivatedEnvironmentVariablesOurselves(
+        resource: Resource,
+        @logValue<PythonEnvironment>('path') interpreter: PythonEnvironment
+    ): Promise<NodeJS.ProcessEnv | undefined> {
+        const workspaceKey = this.workspace.getWorkspaceFolderIdentifier(resource);
+        const key = `${workspaceKey}_${interpreter && getInterpreterHash(interpreter)}`;
+        const shellInfo = defaultShells[this.platform.osType];
+        const envType = interpreter?.envType;
+        if (!shellInfo) {
+            sendTelemetryEvent(Telemetry.GetActivatedEnvironmentVariables, 0, {
+                envType,
+                pythonEnvType: envType,
+                source: 'jupyter',
+                failed: true,
+                reason: 'unknownOS'
+            });
+            return;
+        }
+
+        if (this.activatedEnvVariablesCache.has(key)) {
+            return this.activatedEnvVariablesCache.get(key);
+        }
+
+        const promise = (async () => {
+            const stopWatch = new StopWatch();
+            try {
+                let isPossiblyCondaEnv = false;
+                const [activationCommands, processService] = await Promise.all([
+                    this.getActivationCommands(resource, interpreter),
+                    this.processServiceFactory.create(resource)
+                ]);
+                if (!activationCommands || activationCommands.length === 0) {
+                    sendTelemetryEvent(Telemetry.GetActivatedEnvironmentVariables, stopWatch.elapsedTime, {
+                        envType,
+                        pythonEnvType: envType,
+                        source: 'jupyter',
+                        failed: true,
+                        reason: 'noActivationCommands'
```

Comment:
Better telemetry, this way when things fall over we know what could be causing issues with kernel startups.
E.g. if we cannot activate a conda enviornment, then we know for certain that this is the cause for kernel startup to fail.